Allow setting runtime configuration parameters in the StartupMessage

Brar opened the pull request:
This works either via the `Options` connection string parameter or via
the `PGOPTIONS` environment variable.

As discussed in #2938

There are two remaining tasks:

1. Decide about precedence of parameters set via `Options` vs parameters set directly in the connection string (e. g. "Application Name=App1;Options=application_name=App2").

   In the current implementation the parameters set via `Options` have precedence so the above example would end up with:

   ``` sql
   SHOW application_name;
   
   App2
   ```

2. Documentation
I will add this to the documentation once we've agreed that this is how we want this feature  to be implemented.

roji commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -450,9 +450,74 @@ void WriteStartupMessage(string username)
             if (timezone != null)
                 startupParams["TimeZone"] = timezone;
 
+            var options = Settings.Options?.Length > 0
```
How about moving all the parsing logic to NpgsqlConnectionStringBuilder, and exposing a `Dictionary<string,string>` property there? This would encapsulate all of the connection string parsing logic within that class.

Brar replied:
Yeah I briefly thought about this but than discarded it because it wouldn't work sanely for environment variables that way.
Now that you suggested it again, I thought about it once more and came to the conclusion that, while caching the parsed value from the connection string is good, caching of the parsed value of the environment variable might even be problematic because the user might expect the application to immediately honor changed environment variables, so this is probably even what we want.

I implemented it that way now.

roji replied:
I don't think it's technically possible change environment variables from outside the process once the process has started - at least in Unix, the moment a process is started, they inherit the environment variables from their parent and from that point they have two distinct sets of values. The process can change the environment for themselves, but that's a bit strange.

Anyway, it's not a big deal either way.

Brar replied:
Ok, I've added transparent caching of options from environment variables too.

roji commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -450,9 +450,74 @@ void WriteStartupMessage(string username)
+        static string ParseKey(string str, ref int pos)
+        {
+            var start = pos;
+            for (;pos < str.Length; pos++)
```
nit: space after the semicolon (in ParseValue as well)

roji commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -450,9 +450,74 @@ void WriteStartupMessage(string username)
+                            $"Invalid syntax for connection string parameter '{nameof(NpgsqlConnectionStringBuilder.Options)}': Missing key.");
+                }
+            }
+            throw new FormatException($"Invalid syntax for connection string parameter '{nameof(NpgsqlConnectionStringBuilder.Options)}': Missing '='.");
```
nit: empty line before (in ParseValue as well)

roji commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -450,9 +450,74 @@ void WriteStartupMessage(string username)
+            {
+                if (escaped)
+                    escaped = false;
+                else if(str[pos] == ' ')
```
nit: space space between if and parenthesis

roji commented on src/Npgsql/NpgsqlConnector.cs:
```diff
@@ -450,9 +450,74 @@ void WriteStartupMessage(string username)
+                }
+                else if (str[pos] == '\\')
+                {
+                    escaped = true;
```
Do you really need `escaped` here? You could simply  increment `pos` and fall through to appending it below, simplifying the code (or am I missing something?)

roji commented on test/Npgsql.Tests/ConnectionTests.cs:
```diff
@@ -1353,6 +1367,26 @@ public void Connect_PasswordFromEnvironment_Succeeds()
+            var builder = new NpgsqlConnectionStringBuilder(ConnectionString)
+            {
+                Pooling = false,
+                Options = "default_transaction_isolation=serializable  default_transaction_deferrable=on application_name=My\\ Famous\\\\App"
```
Any reason to assign the options string to builder.Options, and later to set it to null? It can be just set directly into the environment variable (or assigned to some local variable)

Brar commented:
@roji I addressed your feedback.

Next will probably be documentation. 

Which branch shall I do the PR against, https://github.com/npgsql/doc/tree/5.0 or https://github.com/npgsql/doc/tree/release/5.0?

roji commented on src/Npgsql/PostgresEnvironment.cs:
```diff
@@ -22,6 +23,43 @@ static class PostgresEnvironment
+
+        static string? _optionsCache;
+
+        static void ParseOptions()
```
There's a (rare) race condition here if two physical connections connect at the same time, and modify the dictionary concurrently.

To manage this properly, this would either:
* Acquire a lock around the parsing operation, or
* Not operate directly on ParseOptionsCache, but rather on a new dictionary instance in the stack, which would at the end be assigned to ParsedOptionsCache.

FWIW it may not be critical to cache here - establishing a physical connection is quite a heavy thing, and compared to the initial handshake parsing this environment variable is likely to be negligible. So I'm fine with removing this altogether - but if we keep it, it should be safe.

Brar replied:
Yeah, I think so too. I removed the caching of the parsed options. We can always re-enable it if we identify this as a relevant bottleneck or want to squeeze out the last bits of physical connection startup delay for rare configuration variants. ;-)

roji commented on src/Npgsql/PostgresEnvironment.cs:
```diff
@@ -22,6 +23,43 @@ static class PostgresEnvironment
+        {
+            get
+            {
+                if (_optionsCache == Options)
```
Very small nit: if you inverse the logic and the parsing and assignment inside the if block, you get rid of one of the two `return ParseOptionsCache`

Brar replied:
OK. I actually inverted it because it saves me a little indentation but I agree that returning early should probably be reserved for more complex cases.

roji commented:
Let's merge docs to the 5.0 branch for now. If we start publishing actual previews (like .NET/EF Core), it may make sense to make those live, but at the moment these would be docs for a feature that simply isn't available anywhere...

roji commented:
Thanks @Brar!
